### PR TITLE
abilities added :  correlate moduli primes, certificates, and verific…

### PR DESCRIPTION
**Pull request by 7echno**
…ation while running correlation

### CODE REVIEW

File: convert.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import binascii
+import sys
+import hashlib
+
+
```
Comment (tomato42) [inline convert.py]: module docstring missing

File: ecpp [hunk 2]
```diff
@@ -5,14 +5,21 @@ import ConfigParser
...
 # xrange python3 compatibility
 try:
     xrange
 except NameError:
     xrange = range
 
+# input files directory for primo tool 
+primo_in = 'in/'
+certs_dir = 'certificates/'
+#moduli_default = '/etc/ssh/moduli'
```
Comment (tomato42) [inline ecpp]: useless comment

File: convert.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    b = bytearray(howManyBytes)
+    for count in range(howManyBytes-1, -1, -1):
+        b[count] = int(n % 256)
+        n >>= 8
+    return b
+
+def converter(prime_moduli):
+    # use sha256 to hash the prime, then create file with this name with .in end
+    # [Candidate]
+    # N=<hex>
```
Comment (tomato42) [inline convert.py]: that would be better as a docstring (`"""`)

File: ecpp [hunk 4]
```diff
@@ -112,7 +140,102 @@ def main():
...
+        count_line = 1
+        for line in inputfile:
+            if not line.startswith('#'):
+                prime_moduli = (line.split(" ")[6])
+                int_prime_moduli = int(prime_moduli, 16)
+                status = False
+                print("=====================================================")
+                print("Searching cert for: " + str(int_prime_moduli))
+                print("Moduli prime on line no: " + str(count_line))
+                for cert in cert_files:
```
Comment (tomato42) [inline ecpp]: this has quadratic complexity

please load primality certs first and then compare the in-memory list with lines from moduli

File: ecpp [hunk 5]
```diff
@@ -112,7 +140,102 @@ def main():
...
+                                print("[+] cert file " + path + " found for moduli prime on line: " + str(count_line))
+                                status = True
+                                if verify_opt == True:
+                                    verify(path)
+                if status == False:
+                    print("[-] cert file not found for moduli prime")
+                    if primo_opt:
+                        print("[!] creating primo input file in " + primo_in)
+                        converter(prime_moduli)
+            count_line = count_line + 1
```
Comment (tomato42) [inline ecpp]: please use `enumerate()` instead

File: ecpp [hunk 6]
```diff
@@ -112,7 +140,102 @@ def main():
...
+        inputfile = filter(None, (line.rstrip() for line in f))
+        count_line = 1
+        for line in inputfile:
+            if not line.startswith('#'):
+                prime_moduli = (line.split(" ")[6])
+                int_prime_moduli = int(prime_moduli, 16)
+                status = False
+                print("=====================================================")
+                print("Searching cert for: " + str(int_prime_moduli))
+                print("Moduli prime on line no: " + str(count_line))
```
Comment (tomato42) [inline ecpp]: that feels like way too much verbiage for me

a simple
```
line 12: some-file.out
line 13: no certificate found
```
etc would be much more readable

File: ecpp [hunk 7]
```diff
@@ -112,7 +140,102 @@ def main():
...
+
+    for opt, arg in opts:
+        if opt in ('-h', '--help'):
+            usage()
+            sys.exit(2)
+        elif opt in ('-m', '--moduli'):
+            moduli_opt = True
+            moduli = arg
+        elif opt in ('-c', '--certs'):
+            certs_opt = True
```
Comment (tomato42) [inline ecpp]: why set it when `certs` gives the same result? same for `moduli`

File: ecpp [hunk 8]
```diff
@@ -112,7 +140,102 @@ def main():
...
+            usage()
+            sys.exit(2)    
+
+    if not moduli:
+        print("[!] moduli file not specified\n")
+        usage()
+        sys.exit(2)
+
+    if not certs:
+        print("[!] certs directory not specified\n")
```
Comment (tomato42) [inline ecpp]: empty certs are supported by `correlate`, so why require the user to specify it?

File: ecpp [hunk 9]
```diff
@@ -26,8 +33,26 @@ def is_prime_64(n):
...
+def candidate_finder(path):
+    with open(path) as f:
+        config_buff = f.read()
+        f.close()
+        config_ini = StringIO.StringIO(config_buff)
+        config = ConfigParser.RawConfigParser(allow_no_value=True)
+        config.readfp(config_ini)
+        sections = config.sections()
+        v = config.get('PRIMO - Primality Certificate','Format')
+        # print("format version: " + v)
```
Comment (tomato42) [inline ecpp]: unused comment

File: ecpp [hunk 10]
```diff
@@ -26,8 +33,26 @@ def is_prime_64(n):
...
+        if v is "4":
+            n = config.get('Candidate','N')
+            #n = int(n.replace("$",""), 16)
```
Comment (tomato42) [inline ecpp]: useless comment

File: ecpp [hunk 11]
```diff
@@ -112,7 +140,102 @@ def main():
...
+        print("[!] certs directory not specified\n")
+        usage()
+        sys.exit(2)
+    
+    if moduli_opt and not certs_opt and not verify_opt:
+        correlate(moduli)
+    elif moduli_opt and certs_opt and not verify_opt:    
+        correlate(moduli, certs)
+    elif moduli_opt and certs_opt and verify_opt:    
+        correlate(moduli, certs, verify_opt)
```
Comment (tomato42) [inline ecpp]: why not use just this line? the other don't do anything different given how parameters are defined...

File: convert.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    """use sha256 to hash the prime, then create file with this name with .in end
+
+    [Candidate]
+    N=<hex>
+    """
+    n = int(prime_moduli, 16)
+    n = numberToByteArray(n)
+    n = hashlib.sha256(n).hexdigest()
+    file_name = str(n + ".in")
+    with open(primo_in + file_name, "a+") as f:
```
Comment (tomato42) [inline convert.py]: this should be split up to another function

File: ecpp [hunk 13]
```diff
@@ -26,8 +33,24 @@ def is_prime_64(n):
         if n % i == 0:
             raise ValueError("not prime: " + str(n))
 
-def main():
-    with open(sys.argv[1], 'r+') as f:
+def candidate_finder(path):
```
Comment (tomato42) [inline ecpp]: docstring missing

File: ecpp [hunk 14]
```diff
@@ -112,7 +138,95 @@ def main():
     except KeyboardInterrupt:
         print(n)
         raise
-    print("Number is prime")
+    print("\n")
+    print("[+] number is prime\n")
+
+def correlate(moduli, certs=None, verify_opt=None, primo_opt=None):
```
Comment (tomato42) [inline ecpp]: docstring missing

File: ecpp [hunk 15]
```diff
@@ -112,7 +138,95 @@ def main():
...
-    print("Number is prime")
+    print("\n")
+    print("[+] number is prime\n")
+
+def correlate(moduli, certs=None, verify_opt=None, primo_opt=None):    
+    missing_cert_flag = False 
+    cert_files = []
+    if certs is None:
+        certs = certs_dir
+    for path, subdirs, files in os.walk(certs):
```
Comment (tomato42) [inline ecpp]: if a variable is unusued, then we use `_` for its name

File: ecpp [hunk 16]
```diff
@@ -112,7 +138,95 @@ def main():
...
+            cert_files.append(os.path.join(path, name))
+    cert_for_moduli  = []
+    no_cert_for_moduli = []
+    status = None
+    cert_dict = {}
+    for cert in cert_files:
+        if cert.endswith('.out'):
+            path = cert
+            cert_candidate = candidate_finder(path)
+            cert_dict[cert_candidate] = path
```
Comment (tomato42) [inline ecpp]: this whole collection of certs and numbers in them probably would be a good thing to put in a separate function

File: ecpp [hunk 17]
```diff
@@ -112,7 +138,95 @@ def main():
...
+                            path = cert_dict[int_prime_moduli]
+                            print("[+] line {0}: {1}".format(count_line, path))
+                            status = True
+                            if verify_opt == True:
+                                verify(path)
+                if status == False:
+                    print("[-] line {0}: no certificate found".format(count_line))
+                    missing_cert_flag = True
+                    if primo_opt:
+                        print("[+] line {0}: creating input file for {1}".format(count_line,primo_in))
```
Comment (tomato42) [inline ecpp]: space missing in format after comma, also the line is too long, it should not exceed 80 chars

File: ecpp [hunk 18]
```diff
@@ -112,7 +138,95 @@ def main():
...
+    moduli = None
+    moduli_opt = False
+    certs = None
+    certs_opt = False
+    verify = None
+    verify_opt = False
+    primo_opt = False
+
+    try:
+        opts, args = getopt.getopt(sys.argv[1:], 'm:c:pvh', ['moduli=', 'certs=', 'primo=', 'verify=', 'help'])
```
Comment (tomato42) [inline ecpp]: line too long

Comment (tomato42) [inline ecpp]: `primo` option is unused

File: ecpp [hunk 19]
```diff
@@ -112,7 +138,95 @@ def main():
...
+                if status == False:
+                    print("[-] line {0}: no certificate found".format(count_line))
+                    missing_cert_flag = True
+                    if primo_opt:
+                        print("[+] line {0}: creating input file for {1}".format(count_line,primo_in))
+                        converter(prime_moduli)
+    return missing_cert_flag
+
+def usage():
+    print("usage here: ")
```
Comment (tomato42) [inline ecpp]: that's a bit anaemic...

File: ecpp [hunk 20]
```diff
@@ -112,7 +145,99 @@ def main():
...
+                    print("[-] line {0}: no certificate found".format(count_line))
+                    missing_cert_flag = True
+                    if primo_opt:
+                        print("[+] line {0}: creating input file for {1}".format(count_line, primo_in))
+                        create_in_file(primo_in, prime_moduli)
+    return missing_cert_flag
+
+def usage():
+    print("usage here: ")
+
```
Comment (tomato42) [inline ecpp]: usage still missing

File: ecpp [hunk 21]
```diff
@@ -26,8 +33,31 @@ def is_prime_64(n):
...
+    file_name = converter(prime_moduli)
+    file_name = str(file_name + ".in")
+    with open(primo_in + file_name, "w+") as f:
+        f.write("[Candidate]\n")
+        f.write("N=$" + prime_moduli + "\n")
+
+def verify(path):
+    with open(path, 'r+') as f:
         config_buff = f.read()
         f.close()
```
Comment (tomato42) [inline ecpp]: no need to close() when using `with`

File: ecpp [hunk 22]
```diff
@@ -112,7 +145,99 @@ def main():
...
+            cert_dict[cert_candidate] = path
+    with open(str(moduli)) as f:
+        inputfile = filter(None, (line.rstrip() for line in f))
+        for count_line, line in enumerate(inputfile, 1):
+            if not line.startswith('#'):
+                prime_moduli = (line.split(" ")[6])
+                int_prime_moduli = int(prime_moduli, 16)
+                status = False
+                if int_prime_moduli in cert_dict:
+                            path = cert_dict[int_prime_moduli]
```
Comment (tomato42) [inline ecpp]: that indent is a bit too large here

File: ecpp [hunk 23]
```diff
@@ -26,8 +33,31 @@ def is_prime_64(n):
...
+        config_buff = f.read()
+        f.close()
+        config_ini = StringIO.StringIO(config_buff)
+        config = ConfigParser.RawConfigParser(allow_no_value=True)
+        config.readfp(config_ini)
+        sections = config.sections()
+        v = config.get('PRIMO - Primality Certificate','Format')
+        if v is "4":
+            n = config.get('Candidate','N')
+            n = int(n.replace("$",""), 16)
```
Comment (tomato42) [inline ecpp]: space missing after comma

File: convert.py [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+import hashlib
+
+"""
+Functions for hashing the prime number  
+
+"""
+
+def numBits(n):
+    """Return number of bits necessary to represent the integer in binary"""
+    if n==0:
```
Comment (tomato42) [inline convert.py]: space missing around `==`

File: convert.py [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+    if sys.version_info < (2, 7):
+        # bit_length() was introduced in 2.7, and it is an order of magnitude
+        # faster than the below code
+        return len(bin(n))-2
+    else:
+        return n.bit_length()
+
+def numBytes(n):
+    """Return number of bytes necessary to represent the integer in bytes"""
+    if n==0:
```
Comment (tomato42) [inline convert.py]: space missing around `==`

File: ecpp [hunk 26]
```diff
@@ -112,7 +145,99 @@ def main():
...
+
+def main():
+    moduli = None
+    certs = None
+    verify_opt = False
+    primo_opt = False
+    input_file = None
+
+    try:
+        opts, args = getopt.getopt(sys.argv[1:], 'm:c:pvi:h', ['moduli=', 'certs=', 'primo-out', 'verify', 'in=', 'help'])
```
Comment (tomato42) [inline ecpp]: line too long

File: ecpp [hunk 27]
```diff
@@ -112,7 +150,114 @@ def main():
...
+            path = cert
+            cert_candidate = candidate_finder(path)
+            cert_dict[cert_candidate] = path
+    with open(str(moduli)) as f:
+        inputfile = filter(None, (line.rstrip() for line in f))
+        for count_line, line in enumerate(inputfile, 1):
+            if line.startswith('#'):
+                continue
+            prime_moduli = (line.split(" ")[6])
+            int_prime_moduli = int(prime_moduli, 16)
```
Comment (tomato42) [inline ecpp]: the code also should check for `(int_prime_moduli - 1) / 2` to verify that the number is not only prime but a safe prime

File: convert.py [hunk 28]
```diff
@@ -0,0 +1,52 @@
+import binascii
```
Comment (tomato42) [inline convert.py]: please add short licence and copyright

File: ecpp [hunk 29]
```diff
@@ -112,7 +150,114 @@ def main():
...
+                          "creating input file for "
+                          "{1}".format(count_line, primo_in))
+                    create_in_file(primo_in, prime_moduli)
+    return missing_cert_flag
+
+def usage():
+    """Usage"""
+    print("usage: eccp [-h] [-m arg | -i arg] [-c arg] [-p] [-v] \n")
+    print("positional arguments:\n")
+    print("-m, --moduli         specifie moduli file")
```
Comment (tomato42) [inline ecpp]: typo: s/specifie/specify/, also noting that this is the SSH moduli file we're talking about would be a good idea, as well as this is the mode of operation in which the tool will verify that the numbers in moduli file are safe primes, not only that they are prime

File: ecpp [hunk 30]
```diff
@@ -112,7 +150,114 @@ def main():
...
+    print("-c, --certs          specifie certificates dir")
```
Comment (tomato42) [inline ecpp]: typo: s/specifie/specify/

probably also we should qualify that those are primality certificates, ECPP certificates or Primo certificates

File: ecpp [hunk 31]
```diff
@@ -112,7 +150,114 @@ def main():
...
+    print("-p, --primo-out      create primo input file if certificates not found")
```
Comment (tomato42) [inline ecpp]: `primo` is a name, so it should be capitalised

File: ecpp [hunk 32]
```diff
@@ -112,7 +150,114 @@ def main():
...
+    print("-v, --verify         verify prime number for found certificates")
```
Comment (tomato42) [inline ecpp]: technically, we're verifying the certificates, so something like
`verify found primality certificates`

File: ecpp [hunk 33]
```diff
@@ -112,7 +150,114 @@ def main():
...
+    print("-i, --in             verify primo in specified certificate\n")
+
+    print("optional arguments:")
```
Comment (tomato42) [inline ecpp]: aren't most of the arguments optional?

File: ecpp [hunk 34]
```diff
@@ -112,7 +150,114 @@ def main():
...
+
+def usage():
+    """Usage"""
+    print("usage: eccp [-h] [-m arg | -i arg] [-c arg] [-p] [-v] \n")
+    print("positional arguments:\n")
+    print("-m, --moduli         specifie moduli file")
+    print("-c, --certs          specifie certificates dir")
+    print("-p, --primo-out      create primo input file if certificates not found")
+    print("-v, --verify         verify prime number for found certificates")
+    print("-i, --in             verify primo in specified certificate\n")
```
Comment (tomato42) [inline ecpp]: sorry, can't parse this sentence

File: convert.py (hunk 28 above)
Comment (7echno) [inline convert.py]: Something like in tlsfuzzer ? And copy the LICENSE file also ?

# Author: Hubert Kario
# Released under Gnu GPL v2.0, see LICENSE file for details

Though you want to do this.

Comment (tomato42) [inline convert.py]: no, I meant like this: https://github.com/tomato42/ecpp-verifier/blob/master/LICENSE#L460

File: ecpp [hunk 35]
```diff
@@ -39,80 +76,161 @@ def main():
...
+        usage()
+        sys.exit(2)
+
+    if input_file and moduli:
+        print("[!] only certificate or moduli must be specified\n")
+        usage()
+        sys.exit(2)
+
+    if input_file:
+        verify(input_file)
```
Comment (tomato42) [inline ecpp]: so if the verification is successful it won't print anything?

File: ecpp [hunk 36]
```diff
@@ -39,80 +76,161 @@ def main():
...
+
+def usage():
+    """Usage"""
+    print("usage: eccp [-h] [-m arg | -i arg] [-c arg] [-p] [-v] \n")
+    print("arguments:\n")
+    print("-m, --moduli         specify moduli file")
+    print("-c, --certs          specify certificates dir")
+    print("-p, --primo-out      create primo input file if certificates not found")
+    print("-v, --verify         verify found primality certificates")
+    print("-i, --in             verify primality number of a given certificate\n")
```
Comment (tomato42) [inline ecpp]: maybe `verify given primality certificate`?

**Comment by 7echno** [comment]
Should be ready for merge now. I suggest to go step by step for implementing anything new.

File: ecpp [hunk 37]
```diff
@@ -1,35 +1,75 @@
...
 
-def main():
-    with open(sys.argv[1], 'r+') as f:
+def candidate_finder(path):
+    """Return the int number from certificates Candidate section.
+       If file is malformed, file is not primality certificate, or version
+       of primality certificate is unsupported 'None' will be returned"""
+    with open(path) as f:
         config_buff = f.read()
         f.close()
```
Comment (tomato42) [inline ecpp]: when `with open()` is used, then `f.close()` is not needed

also, if the contents of the file are in an array, then it's not necessary to put all method code inside the `with` section

File: convert.py [hunk 38]
```diff
@@ -0,0 +1,55 @@
...
+        return len(bin(n))-2
+    else:
+        return n.bit_length()
+
+def numBytes(n):
+    """Return number of bytes necessary to represent the integer in bytes"""
+    if n == 0:
+        return 0
+    bits = numBits(n)
+    return (bits + 7) // 8
```
Comment (tomato42) [inline convert.py]: when using `//` it's customary to do `from __future__ import division`

File: convert.py [hunk 39]
```diff
@@ -0,0 +1,55 @@
...
+    The returned bytearray may be smaller than howManyBytes, but will
+    not be larger.  The returned bytearray will contain a big-endian
+    encoding of the input integer (n).
+    """
+    if howManyBytes == None:
+        howManyBytes = numBytes(n)
+    b = bytearray(howManyBytes)
+    for count in range(howManyBytes-1, -1, -1):
+        b[count] = int(n % 256)
+        n >>= 8
```
Comment (tomato42) [inline convert.py]: current code in tlslite-ng is much faster and cleaner:
```
return bytearray((n >> i) & 0xff                                        
                         for i in reversed(range(0, howManyBytes*8, 8)))
```

File: convert.py [hunk 40]
```diff
@@ -0,0 +1,55 @@
...
+    return b
+
+def converter(int_prime_moduli):
```
Comment (tomato42) [inline convert.py]: how about `number_as_hash` or `sha256_of_number`?

File: ecpp [hunk 41]
```diff
@@ -1,35 +1,75 @@
...
 
+# Author: Hubert Kario, Stefan Dordevic
+# Released under Gnu GPL v2.1, see LICENSE file for details
+
 from __future__ import print_function
 import ConfigParser
 import sys
 import StringIO
 from verify import *
+from convert import *
```
Comment (tomato42) [inline ecpp]: star imports are not recommended, they make it harder for static analysers to detect unused code

File: ecpp [hunk 42]
```diff
@@ -1,35 +1,75 @@
...
 import getopt
+import os
+import re
```
Comment (tomato42) [inline ecpp]: system imports should be placed before local imports

File: ecpp [hunk 43]
```diff
@@ -1,35 +1,75 @@
...
+# input files directory for primo tool
+PRIMO_IN = 'in/'
+CERTS_DIR = 'certificates/'
+
 # dict with params
-params = {}
+PARAMS = {}
 def par_set():
-    params.update({'s':None, 'w':None, 'j':None, 't':None, 'a':None, 'b':None, 'q':None, 'r':None})
+    """Paramiter set/reset function"""
```
Comment (tomato42) [inline ecpp]: `s/Paramiter/Parameter/`

File: ecpp [hunk 44]
```diff
@@ -1,35 +1,75 @@
...
+    PARAMS.update({'s':None, 'w':None, 'j':None, 't':None,
+                   'a':None, 'b':None, 'q':None, 'r':None})
 
 # division function
 def is_prime_64(n):
+    """Check if number is a prime through division function"""
```
Comment (tomato42) [inline ecpp]: "through trial division"

File: ecpp [hunk 45]
```diff
@@ -1,35 +1,75 @@
...
+def create_in_file(directory, int_prime_moduli):
+    """Create input file for primo tool"""
+    file_name = converter(int_prime_moduli)
+    file_name = str(file_name + ".in")
+    with open(directory + file_name, "w+") as f:
+        f.write("[Candidate]\n")
+        f.write("N={0}\n".format(int_prime_moduli))
+
+def verify(path):
+    """Prime number verification"""
```
Comment (tomato42) [inline ecpp]: isn't it primality certificate verification?

File: ecpp (hunk 42 above)
Comment (tomato42) [inline ecpp]: `re` is imported but not used

File: ecpp [hunk 46]
```diff
@@ -1,35 +1,75 @@
...
+    """Return the int number from certificates Candidate section.
+       If file is malformed, file is not primality certificate, or version
+       of primality certificate is unsupported 'None' will be returned"""
+    with open(path) as f:
         config_buff = f.read()
         f.close()
+        config_ini = StringIO.StringIO(config_buff)
+        config = ConfigParser.RawConfigParser(allow_no_value=True)
+        config.readfp(config_ini)
+        sections = config.sections()
```
Comment (tomato42) [inline ecpp]: `sections` unused

File: ecpp [hunk 47]
```diff
@@ -1,35 +1,75 @@
...
+CERTS_DIR = 'certificates/'
+
 # dict with params
-params = {}
+PARAMS = {}
 def par_set():
-    params.update({'s':None, 'w':None, 'j':None, 't':None, 'a':None, 'b':None, 'q':None, 'r':None})
+    """Paramiter set/reset function"""
+    PARAMS.update({'s':None, 'w':None, 'j':None, 't':None,
+                   'a':None, 'b':None, 'q':None, 'r':None})
```
Comment (tomato42) [inline ecpp]: nit: missing whitespace after `:`

File: ecpp [hunk 48]
```diff
@@ -1,35 +1,75 @@
...
 except NameError:
     xrange = range
 
+# input files directory for primo tool
+PRIMO_IN = 'in/'
+CERTS_DIR = 'certificates/'
+
 # dict with params
-params = {}
+PARAMS = {}
```
Comment (tomato42) [inline ecpp]: nit: top-level funcion and class definitions should have 2 blank lines before their definition

File: ecpp [hunk 49]
```diff
@@ -39,80 +79,165 @@ def main():
...
+            primes.append(sophie_prime_moduli)
+            print_out = []
+            for p in primes:
+                if p in cert_dict:
+                    print_out.append("cert: {0}".format(path))
+                    if verify_opt:
+                        if verify(cert_dict[p]):
+                            print_out[0] = ("cert: {0} - is a prime number".format(path))
+                        else:
+                            print_out[0] = ("cert: {0} - is not a prime number".format(path))
```
Comment (tomato42) [inline ecpp]: this and line 159: too long (over 79 characters)

File: ecpp [hunk 50]
```diff
@@ -39,80 +79,165 @@ def main():
...
+            cert_candidate = candidate_finder(path)
+            cert_dict[cert_candidate] = path
+    with open(str(moduli)) as f:
+        inputfile = filter(None, (line.rstrip() for line in f))
+        for count_line, line in enumerate(inputfile, 1):
+            if line.startswith('#'):
+                continue
+            prime_moduli = (line.split(" ")[6])
+            int_prime_moduli = int(prime_moduli, 16)
+            sophie_prime_moduli = (int_prime_moduli - 1) / 2
```
Comment (tomato42) [inline ecpp]: I think you wanted to use `//` here

File: ecpp [hunk 51]
```diff
@@ -39,80 +79,165 @@ def main():
...
+                          ", ".join(print_out)))
+    return missing_cert_flag
+
+def usage():
+    """Usage"""
+    print("usage: eccp [-h] [-m arg | -i arg] [-c arg] [-p] [-v] \n")
+    print("arguments:\n")
+    print("-m, --moduli         specify moduli file")
+    print("-c, --certs          specify certificates dir")
+    print("-p, --primo-out      create primo input file if certificates not found")
```
Comment (tomato42) [inline ecpp]: line too long

File: ecpp [hunk 52]
```diff
@@ -39,80 +79,165 @@ def main():
...
+def main():
+    """Main function"""
+    moduli = None
+    certs = None
+    verify_opt = False
+    primo_opt = False
+    input_file = None
+
+    try:
+        opts, args = getopt.getopt(sys.argv[1:], 'm:c:pvi:h', ['moduli=', 'certs=',
```
Comment (tomato42) [inline ecpp]: line too long

File: convert.py (hunk 28 above)
Comment (tomato42) [inline convert.py]: also, `binascii` is imported but not used

**Comment by 7echno** [comment]
All of above should be fixed now.

**Comment by tomato42** [comment]
looks good, thanks!

**Pull request merged**